src/datafusion/table_provider.rs

```diff
@@ -0,0 +1,136 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use arrow_schema::{Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    common::Statistics,
+    datasource::{
+        listing::{ListingTableUrl, PartitionedFile},
+        physical_plan::FileScanConfig,
+        TableProvider, TableType,
+    },
+    execution::context::SessionState,
+    logical_expr::{Expr, TableProviderFilterPushDown},
+    physical_plan::ExecutionPlan,
+};
+
+use crate::{
+    async_reader::{ZarrPath, ZarrReadAsync},
+    reader::ZarrResult,
+};
+
+use super::scanner::ZarrScan;
+
+pub struct ListingZarrTableOptions {}
+
+impl ListingZarrTableOptions {
+    pub async fn infer_schema(
+        &self,
+        state: &SessionState,
+        table_path: &ListingTableUrl,
+    ) -> ZarrResult<Schema> {
+        let store = state.runtime_env().object_store(table_path)?;
+
+        let zarr_path = ZarrPath::new(store, table_path.prefix().clone());
+        let schema = zarr_path.get_zarr_metadata().await?.arrow_schema()?;
+
+        Ok(schema)
+    }
+}
+
+pub struct ListingZarrTableConfig {
+    /// The inner listing table configuration
+    table_path: ListingTableUrl,
+}
+
+impl ListingZarrTableConfig {
+    /// Create a new ListingZarrTableConfig
+    pub fn new(table_path: ListingTableUrl) -> Self {
+        Self { table_path }
+    }
+}
+
+pub struct ZarrTableProvider {
+    table_schema: Schema,
+    config: ListingZarrTableConfig,
+}
+
+impl ZarrTableProvider {
+    pub fn new(config: ListingZarrTableConfig, table_schema: Schema) -> Self {
+        Self {
+            table_schema,
+            config,
+        }
+    }
+}
+
+#[async_trait]
+impl TableProvider for ZarrTableProvider {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::new(self.table_schema.clone())
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> datafusion::error::Result<Vec<TableProviderFilterPushDown>> {
+        // TODO: which filters can we push down?
```

Comment:
This is how datafusion knows which filters can be pushed down. See the `TableProviderFilterPushDown` in datafusion for options. 

Defaulting everything to `Unsupported` as the default. But I have an example of something here that intercepts a UDF: https://github.com/wheretrue/exon/blob/0bf13bbfb80e79fbaedab52cbccda2ec62d4fd22/exon/exon-core/src/datasources/vcf/table_provider.rs#L271-L292.

Reply:
A filter here is any condition on one or more columns in the store right? That's basically already supported, we can push filters down to the reader or stream and it will only return the rows that satisfy the condition(s), while avoiding to read whole chunks if possible (that part is kind of a trade off, happy to discuss how I did that at some point).

So as you can see from my comment on the issue for the async stream and blocking tasks, I'm going to commit something and close that issue soon (more details in my comment). I plan on starting to work more on the datafusion implementation, filters might be a good place for me to start? I'll create an issue for it once we merge this and we can discuss there.

Reply:
Yeah, if you wanna add the predicate pushdown integration between datafusion and the underlying reader, that makes sense as a good place to start for me.

This is the function that determines how the query plan will behave, i.e. if the predicate is supplied to `scan` as `filters`, which you can see is unused in this PR, or the predicate remains a separate node.

Reply:
Sounds good, I'll start on this within a few days I think.